collections/quad.go

```diff
@@ -0,0 +1,425 @@
+	err := json.Unmarshal(b, &jsonKey)
+	if err != nil {
+		return Quad[K1, K2, K3, K4]{}, err
+	}
+
+	key1, err := t.keyCodec1.DecodeJSON(jsonKey[0])
+	if err != nil {
+		return Quad[K1, K2, K3, K4]{}, err
+	}
+
+	key2, err := t.keyCodec2.DecodeJSON(jsonKey[1])
+	if err != nil {
+		return Quad[K1, K2, K3, K4]{}, err
+	}
+
+	key3, err := t.keyCodec3.DecodeJSON(jsonKey[2])
+	if err != nil {
+		return Quad[K1, K2, K3, K4]{}, err
+	}
+
+	key4, err := t.keyCodec4.DecodeJSON(jsonKey[3])
+	if err != nil {
+		return Quad[K1, K2, K3, K4]{}, err
+	}
+
+	return Join4(key1, key2, key3, key4), nil
+}
+
+// Stringify converts Quads to string
+func (t quadKeyCodec[K1, K2, K3, K4]) Stringify(key Quad[K1, K2, K3, K4]) string {
+	b := new(strings.Builder)
+	b.WriteByte('(')
+	if key.k1 != nil {
+		b.WriteByte('"')
+		b.WriteString(t.keyCodec1.Stringify(*key.k1))
+		b.WriteByte('"')
+	} else {
+		b.WriteString("<nil>")
+	}
+
+	b.WriteString(", ")
+	if key.k2 != nil {
+		b.WriteByte('"')
+		b.WriteString(t.keyCodec2.Stringify(*key.k2))
+		b.WriteByte('"')
+	} else {
+		b.WriteString("<nil>")
+	}
+
+	b.WriteString(", ")
+	if key.k3 != nil {
+		b.WriteByte('"')
+		b.WriteString(t.keyCodec3.Stringify(*key.k3))
+		b.WriteByte('"')
+	} else {
+		b.WriteString("<nil>")
+	}
+
+	b.WriteString(", ")
+	if key.k4 != nil {
+		b.WriteByte('"')
+		b.WriteString(t.keyCodec4.Stringify(*key.k4))
+		b.WriteByte('"')
+	} else {
+		b.WriteString("<nil>")
+	}
+
+	b.WriteByte(')')
+	return b.String()
+}
+
+func (t quadKeyCodec[K1, K2, K3, K4]) KeyType() string {
+	return fmt.Sprintf("Quad[%s,%s,%s,%s]", t.keyCodec1.KeyType(), t.keyCodec2.KeyType(), t.keyCodec3.KeyType(), t.keyCodec4.KeyType())
+}
+
+func (t quadKeyCodec[K1, K2, K3, K4]) Encode(buffer []byte, key Quad[K1, K2, K3, K4]) (int, error) {
+	writtenTotal := 0
+	if key.k1 != nil {
+		written, err := t.keyCodec1.EncodeNonTerminal(buffer, *key.k1)
+		if err != nil {
+			return 0, err
+		}
+		writtenTotal += written
+	}
+	if key.k2 != nil {
+		written, err := t.keyCodec2.EncodeNonTerminal(buffer[writtenTotal:], *key.k2)
+		if err != nil {
+			return 0, err
+		}
+		writtenTotal += written
+	}
+	if key.k3 != nil {
+		written, err := t.keyCodec3.EncodeNonTerminal(buffer[writtenTotal:], *key.k3)
+		if err != nil {
+			return 0, err
+		}
+		writtenTotal += written
+	}
+	if key.k4 != nil {
+		written, err := t.keyCodec4.Encode(buffer[writtenTotal:], *key.k4)
+		if err != nil {
+			return 0, err
+		}
+		writtenTotal += written
+	}
+	return writtenTotal, nil
+}
+
+func (t quadKeyCodec[K1, K2, K3, K4]) Decode(buffer []byte) (int, Quad[K1, K2, K3, K4], error) {
+	readTotal := 0
+	read, key1, err := t.keyCodec1.DecodeNonTerminal(buffer)
+	if err != nil {
+		return 0, Quad[K1, K2, K3, K4]{}, err
+	}
+	readTotal += read
+	read, key2, err := t.keyCodec2.DecodeNonTerminal(buffer[readTotal:])
+	if err != nil {
+		return 0, Quad[K1, K2, K3, K4]{}, err
+	}
+	readTotal += read
+	read, key3, err := t.keyCodec3.DecodeNonTerminal(buffer[readTotal:])
```

Comment:
Similar as previous comment. Also k4 seems to be using the good function
```suggestion
	read, key1, err := t.keyCodec1.Decodel(buffer)
	if err != nil {
		return 0, Quad[K1, K2, K3, K4]{}, err
	}
	readTotal += read
	read, key2, err := t.keyCodec2.Decode(buffer[readTotal:])
	if err != nil {
		return 0, Quad[K1, K2, K3, K4]{}, err
	}
	readTotal += read
	read, key3, err := t.keyCodec3.Decode(buffer[readTotal:])
```